Cohort filters (solves #952)

Twixes opened the pull request:
## Changes

Added filtering by cohort, as described in #952. Here's the front end:

![Cohort filter preview](https://user-images.githubusercontent.com/4550621/86249452-9b749080-bbaf-11ea-9872-694cd166c193.png)

And on the back end, this is a new type of property:
```TypeScript
interface CohortProperty {
	key: 'id',
	value: number, // cohort ID, e.g. 42
	type: 'cohort'
}
```

## Checklist
- [x] All querysets/queries filter by Team (if applicable)
- [ ] Backend tests (if applicable)
- [ ] Cypress E2E tests (if applicable)

Twixes committed:
Update UI for cohort filters

Twixes committed:
Change cohort filter interface slightly

In TS notation a cohort filter is: `{
    key: string, // hashtag followed by cohort ID
    value: true,
    type: 'cohort'
}`

Twixes committed:
Add server-side support for cohort filters

Twixes committed:
Optimize PropertyFilter and change tabs width

Twixes commented:
Draft for now as I'm looking into tests and if this doesn't cause any issue anywhere. Might be good to refactor the JSX too, PropertyFilter was already big before and only got bigger.

Twixes committed:
Refactor PropertyFilter for better maintainability

Twixes committed:
Fix property of undefined bug

Twixes committed:
Fix Select values

Twixes committed:
Test API

timgl commented:
Somne merge conflicts, happy to take a look once those are solved :)

Twixes committed:
Resolve merge conflicts with master

Twixes committed:
Prettier JS

Twixes committed:
Merge branch 'master' into 952-filter-by-cohort

timgl commented on frontend/src/lib/components/PropertyFilters/PropertyFilter.js:
```diff
@@ -110,24 +125,105 @@ export function PropertyFilter({ index, onComplete, logic }) {
+}
+
+function CohortPaneContents({ onComplete, setThisFilter, propkey, displayOperatorAndValue }) {
+    const { cohorts } = useValues(cohortsModel)
```
`BreakdownFilter.js` also has a Cohort filter, it could be worth extracting that into a component and using that (though in this case without "* All users"). That way you'll also be able to search this Select, which you can't do at the moment

Twixes replied:
It does have a cohort filter but for selecting multiple filters, so a bit of a different use case. Definitely should be searchable though – just made it so.

timgl replied:
Yep you're absolutely right, good catch

timgl commented on frontend/src/lib/components/PropertyFilters/PropertyFilter.js:
```diff
@@ -110,24 +125,105 @@ export function PropertyFilter({ index, onComplete, logic }) {
+                {cohorts.map((item, index) => (
+                    <Select.Option
+                        key={'cohort_' + item.id}
+                        value={`#${item.id}`}
```
This is a little weird and feels a little brittle. Wouldn't it be easier to just have a fixed key `cohort_id`, then make the value the id of the cohort?

Twixes replied:
I think it indeed would. Reworked this type so that it's
```TypeScript
interface CohortProperty {
	key: 'id', // was: string (hashtag followed by cohort ID, e.g. '#42'), now: 'id'
	value: number, // was: true, now: number (cohort ID, e.g. 42)
	type: 'cohort'
}
```
both front and back end.

Twixes committed:
Add search to CohortPaneContents and element attributes

Twixes committed:
Make cohort filter property more elegant

The relevant type is now
```TypeScript
interface CohortProperty {
	key: 'id',
	value: number, // cohort ID, e.g. 42
	type: 'cohort'
}
```

Twixes committed:
Update and enhance test

Twixes committed:
Improve code style
